kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/NettyFilterDispatchExecutorTest.java

```diff
@@ -111,7 +126,68 @@ void completeOnFilterDispatchThreadWhenInEventLoop() {
             CompletionStage<Boolean> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage)
                     .thenApply(o -> dispatchExecutor.isInFilterDispatchThread());
             completionStage.complete("arbitrary");
-            assertThat(inDispatchThread).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+            assertThat(toStandardCompletableFuture(inDispatchThread)).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    static <T> CompletableFuture<T> toStandardCompletableFuture(CompletionStage<T> stage) {
+        CompletableFuture<T> future = new CompletableFuture<>();
+        stage.whenComplete((o, throwable) -> {
+            if (throwable != null) {
+                future.completeExceptionally(throwable);
+            }
+            else {
+                future.complete(o);
+            }
+        });
+        return future;
+    }
+
+    @Test
+    void completeOnFilterDispatchThreadResultCannotBeConvertedToCompletableFuture() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Object> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage);
+            assertThatThrownBy(inDispatchThread::toCompletableFuture)
+                    .isInstanceOf(UnsupportedOperationException.class)
+                    .hasMessage("CompletableFuture usage disallowed, we don't want to block the event loop or allow unexpected completion");
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    @Test
+    void completeOnFilterDispatchThreadResultNotCompletableByClientByCasting() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Object> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage);
+            boolean isCompletableFuture = inDispatchThread instanceof CompletableFuture;
+            // cannot assertThat(inDispatchThread) because it attempts to use toCompletableFuture immediately
+            assertThat(isCompletableFuture).isFalse();
```

Comment:
not sure if `new ObjectAssert<>(inDispatchThread).isNotInstanceOf(CompletableFuture.class);` is a better or worse alternative.

Reply:
Thanks, that lead me to `Assertions.assertThatObject`